src/editor/editor_brush_ui.cpp

```diff
@@ -0,0 +1,418 @@
+
+	manualEditMode = std::make_unique<gcn::CheckBox>("Manual mode");
+	manualEditMode->setHeight(14);
+	manualEditMode->setFont(&GetGameFont());
+	parrent->add(manualEditMode.get());
+	manualEditMode->setPosition(UIRectangle.x + 5,
+				 				controlSets[ECtrlSets::cSize].bottomRight().y + verticalGap);
+
+	controlSets[ECtrlSets::cSingleTile].add(dynamic_cast<gcn::Widget*>(manualEditMode.get()));
+
+	manualEditModeListener = std::make_unique<LambdaActionListener>([this](const std::string &) {
+
+		auto &brush = Editor.brushes.getCurrentBrush();
+		brush.enableFixNeighbors(manualEditMode->isSelected() == false);
+		enableRnd->setVisible(brush.isRandomizeAllowed() && manualEditMode->isSelected());
+
+		if (manualEditMode->isSelected() == false) {
+			brush.setDecorative(false);
+		}
+		decorative->setVisible(!brush.isFixNeighborsEnabled());
+		decorative->setSelected(brush.isDecorative());
+
+		Editor.tileIcons.rebuild(brush.isFixNeighborsEnabled() == false,
+								 brush.isRandomizationEnabled());
+	});
+	manualEditMode->addActionListener(manualEditModeListener.get());
+
+	enableRnd = std::make_unique<gcn::CheckBox>("Random");
+	enableRnd->setHeight(14);
+	enableRnd->setFont(&GetGameFont());
+	parrent->add(enableRnd.get());
+	enableRnd->setPosition(manualEditMode->getX(),
+				 			controlSets[ECtrlSets::cSingleTile].bottomRight().y + verticalGap);
+	controlSets[ECtrlSets::cSingleTile].add(dynamic_cast<gcn::Widget*>(enableRnd.get()));
+
+	enableRndListener =
+	std::make_unique<LambdaActionListener>([this](const std::string&) {
+
+		auto &brush = Editor.brushes.getCurrentBrush();
+		brush.enableRandomization(enableRnd->isSelected());
+		Editor.tileIcons.rebuild(brush.isFixNeighborsEnabled() == false,
+								 brush.isRandomizationEnabled());
+	});
+	enableRnd->addActionListener(enableRndListener.get());
+
+	decorative = std::make_unique<gcn::CheckBox>("Decorative");
+	decorative->setHeight(14);
+	decorative->setFont(&GetGameFont());
+
+	parrent->add(decorative.get());
+	decorative->setPosition(manualEditMode->getX(),
+				 			controlSets[ECtrlSets::cSingleTile].bottomRight().y + verticalGap);
+	controlSets[ECtrlSets::cSingleTile].add(dynamic_cast<gcn::Widget*>(decorative.get()));
+
+	decorativeListener = std::make_unique<LambdaActionListener>([this](const std::string &) {
+		auto &brush = Editor.brushes.getCurrentBrush();
+		brush.setDecorative(decorative->isSelected());
+	});
+	decorative->addActionListener(decorativeListener.get());
+
+	reloadCtrlSettings();
+}
+
+void CBrushControlsUI::reloadCtrlSettings()
+{
+	for (auto &[key, ctrl] : controlSets) {
+		ctrl.resetHidden();
+	}
+	const auto brush = Editor.brushes.getCurrentBrush();
+	Editor.tileIcons.enable(brush.isTileIconsPaletteRequired());
+	updateSingleTileCtrls();
+	updateSizeCtrls();
+	updateGeneratorOptionsCtrls();
+}
+
+void CBrushControlsUI::updateSingleTileCtrls()
+{
+	const auto brush = Editor.brushes.getCurrentBrush();
+	CBrush &_brush = Editor.brushes.getCurrentBrush();
```

Comment:
Why both? should be `const auto&`?

Reply:
fixed:
https://github.com/Wargus/stratagus/pull/732/commits/ae8997e651e5bf786e2164921889500a25329e93